### ISSUE TITLE
Bug in the flux-calibration variance calculation ? 

### ISSUE THREAD

**Issue by segasai** [desihub/desispec#2420]
Hi, 

While preparing for #2419, 
I've looked into fluxcalibration code and I believe the the current code that compute the ivar's of flux-calibrated frames is not quite correct:

https://github.com/desihub/desispec/blob/255647f0d2b7038c4747d9abb86af49bcceea9d6/py/desispec/fluxcalibration.py#L1478

Specifically the current code uses this eqn:

$$Newivar =  1. / ( 1. /(iVar(F) C^2) + F^2 /(iVar(C) C^4))$$
where Ivar(C) is the inverse variance of fluxcalib vector, and  ivar(F) is the inverse variance of spectra.
The problem here, is that it applies this eqn to already calibrated spectra (i..e original flux divided by C) 
see https://github.com/desihub/desispec/blob/255647f0d2b7038c4747d9abb86af49bcceea9d6/py/desispec/fluxcalibration.py#L1473

While the equation only applies to original undivided flux. 

Derivation:

$$F1 = F/C$$

$$Var(F1) = 1/C^2 Var(F) + F^2  Var(1/C) = 1/C^2 Var(F) + F^2 /C^4  Var(C)$$

$$iVar(F1) = 1./(1/C^2 Var(F) + F^2 /C^4 Var(C)) = 1./(1/(C^2 iVar(F)) + F^2 /(C^4  iVar(C)) )  $$
This is exactly what the current code uses, but instead of F in it, it uses F1.
Therefore this leads to incorrect C^2 factor in the second term.
Essentially the current code does: 
$$iVar(F1) = 1./(1/C^2 Var(F) + F^2 /C^4 Var(C)) = 1./(1/(C^2 iVar(F)) + F^2 /(C^6  IVar(C)) )  $$
A quick check shows the typical value of C is ~20-70, so that will lower the second term significantly, reducing the effect from flux-calibration error.

I am expecting that this will affect variances, where flux calibration uncertainty starts to become comparable to the spectrum uncertainty.
I don't know a typical S/N of flux calibration vectors a look at a few random files they can be pretty low, i.e. 20.

Am I missing something here ? 
        S

**Comment by segasai** [comment]
Assuming that I don't do something stupid here, the effect from this bug is massive I believe.
I.e. I compared the reduction with the #2422 fix and loa, and here's a S/N new vs S/N original
![image](https://github.com/user-attachments/assets/308502d9-87ca-4ea0-889a-18af7166d8e4)
This is from /global/cfs/cdirs/desi/spectro/redux/koposov/fluxcalib_fix/exposures/

I.e. we are talking about factor of two decrease of S/N for high S/N pixels.

**Comment by segasai** [comment]
And here's the confirmation that the effect is real.
I took spectra- file (spectra-sv3-bright-15354.fits.gz from kibo )  and looked at the $\delta=\frac{S_1-S_2}{\sqrt{E_1^2+E_2^2}}$ for pairs of spectra vs S/N and one can see clearly that at high S/N the distributions broadens away from N(0,1), matching what's expected based on reasoning above.
![image](https://github.com/user-attachments/assets/65033016-d301-4779-b03e-acbcd7b9ee34)

**Comment by segasai** [comment]
For stars, our model spectra are "always" wrong, so I personally always ignored the increase in chi-square at high S/N, therefore I didn't notice this earlier. 
But IMO the information content would change drastically with this and I believe it's relevant at S/N much lower than 100 (i.e. your 1% uncertainty floor). I.e. the example plot from above shows that S/N=20 could be S/N=10 (depending on the S/N of fluxcalib).

**Comment by julienguy** [comment]
I am afraid you are correct @segasai . The bug results in underestimating the contribution of the flux calibration statistical error to the calibrated flux variance. It affects the uncertainties of bright spectra. This is why we did not detect this so far in Lya P1D (where the noise was validated down to 1% of variance).

**Comment by araichoor** [comment]
kind of vaguely related: in Jan. 2023, I did look at sky fibers (in rosette-like observations of tertiary programs) and checked that the `flux * sqrt(ivar)` distributions where close to a N(0, 1) distributions; ie confirming that this bug shouldn t affect the low snr spectra.
here, and subsequent posts: https://desisurvey.slack.com/archives/C0351RV8CBE/p1673051607355669.

**Comment by segasai** [comment]
Yes, thanks, @araichoor , in MWS I also looked at low S/N often and I've been tracking the chi-square close to 1 there, so I think it's clear that low S/N is okay. But I'd expect to see differences from S/N=10-20. I don't have a good intuition how much higher the S/N in fluxcalib is in dark survey if that indeed leads to issues only at higher S/Ns -- that would explain the Lya results @julienguy mentioned.

**Comment by segasai** [comment]
BTW
I've checked the S/N of fluxcalib in the dark program, and across random set of exposures, I'm seeing median values of 20 (with tail to 40). That means dark program will equally experience this issue and a typical DESI pixel with S/N=20 has uncertainties overestimated by ~ 40%. That corresponds to roughly 5% error floor added in quadrature ( for a single exposure, but that will be reduced by sqrt(N) from stacking multiple exposures).

**Comment by segasai** [comment]
On this issue (in the context of the tertiary programs).
What would be the best way of creating the new cframes for a list of expid's from loa with this fix in place. I.e. if I want to recreate the coadds for various tertiaries.

In the past I just did desi_group_spectra, desi_coadd_spectra using cframes from the selected expids as inputs.
But now I also want to recreate the cframes first.

Could anyone suggest the stanza for that, or point me somewhere. Presumably that would involve desi_compute_fluxcalibration, but I don't know all the flags that needs go there. 
Thanks in advance.

**Comment by segasai** [comment]
Thanks @sbailey . 
Interesting... 
It'd be worthwhile to probably have independent check of correctness of ivar's by looking at the repeated observations. This would gave us more confidence that it was incorrect before #2496 , and is correct now.

**Comment by sbailey** [comment]
@segasai Good idea on the repeated observations. Could you do that and post results?  #2496 included an analysis of the sigma of the (flux-model)/err of stdstar fits (while including a broadband warping term), but a direct comparison of data:data would also be great.

**Comment by segasai** [comment]
@sbailey I have hacked a script that does check that.
I have selected a tile (tile 70) that seems to have enough bright stars and repeats.

And here're are the plots of the 
log10(Avg (SNR)) vs $Stddev[(S_i - S_j)/\sqrt(1/Ivar_i + 1/Ivar_j)]$
Where S_i, S_j, Ivar_i, Ivar_j refer to i-th observation/ j-th observation of a pixel
if the errors are perfect the values should be close to 1. 

3 plots correspond to B/R/Z arms. And the plot is for loa.

![Image](https://github.com/user-attachments/assets/0921beb6-da98-4127-971d-bade6a2f62d1)

![Image](https://github.com/user-attachments/assets/e304d8eb-4000-4f9a-ad1c-e7a57381ca5a)


![Image](https://github.com/user-attachments/assets/6a9937c2-e18b-4bc4-8618-b51f9873a440)


The thing I don't know how to do easily (to test the new code that was committed) is how to make the spectra-* files for a all observations of a given tile (without redoing the spectral extractions).

The script I used is below (should be runnable on NERSC):

```
import glob
import numpy as np
import astropy.table as atpy
import astropy.io.fits as pyfits
import duckdb
import matplotlib.pyplot as plt
import scipy.stats

Ds = []
D_IVs = []
D_Ms = []
Ts = []
arm = 'Z'
fdir = '/global/cfs/cdirs/desi/spectro/redux/loa/tiles/cumulative/70/20210430/'
fdir = 'data/'
# from https://data.desi.lbl.gov/desi/spectro/redux/loa/tiles/cumulative/70/20210430/spectra-$a-70-thru20210430.fits.gz

for f in glob.glob('data/*spec*fits*'):
    D = pyfits.getdata(f, f'{arm}_FLUX')
    D_IV = pyfits.getdata(f, f'{arm}_IVAR')
    D_M = pyfits.getdata(f, f'{arm}_MASK')
    T = atpy.Table().read(f, 'FIBERMAP')
    Ds.append(D)
    D_IVs.append(D_IV)
    Ts.append(T)
    D_Ms.append(D_M)

Ts = atpy.vstack(Ts)
Ds = np.concatenate(Ds, axis=0)
D_IVs = np.concatenate(D_IVs, axis=0)
D_Ms = np.concatenate(D_Ms, axis=0)

Ts['ROWID'] = np.arange(len(Ts))

C = duckdb.connect(':memory:')
C.register('tt', Ts.to_pandas())
xid1, xid2 = list(
    C.execute('''select t1.rowid,t2.rowid from tt as t1, tt as t2
where t1.rowid!=t2.rowid and t1.targetid=t2.targetid''').fetchnumpy().values())
(Dm1, Dm2, D1, D2, Div1, Div2) = (D_Ms[xid1], D_Ms[xid2], Ds[xid1], Ds[xid2],
                                  D_IVs[xid1], D_IVs[xid2])
xgrid = np.linspace(-10, 10, 1000)
plt.clf()
xind0 = (Div2 != 0) & (Dm1 == 0) & (Dm2 == 0) & (Div1 != 0)
xind = (Div2 != 0) & (Dm1 == 0) & (Dm2 == 0) & (Div1 != 0) & ((
    (D1 + D2) * .5 / (0.5 / Div1 + 0.5 / Div2)**.5) > 35)
binr = [-10, 10]
nbins = 100
binsize = (binr[1] - binr[0]) / nbins
plt.hist(((D1 - np.nanmedian(D1 / D2, axis=1)[:, None] * D2) /
          ((1 / Div1 + 1 / Div2)**.5))[xind],
         range=binr,
         bins=nbins,
         histtype='step')
plt.plot(xgrid,
         scipy.stats.norm(0, 1).pdf(xgrid) * xind.sum() * binsize,
         color='red')
thresh = 0.1
X = (np.log10((D1 + D2) * .5 / (0.5 / Div1 + 0.5 / Div2)**.5))
scale = (np.nanmedian(D1, axis=1) / (np.nanmedian(D2, axis=1)))[:, None]
Y = ((D1 - scale * D2) / ((1 / Div1 + 1**2 / Div2)**.5))
xind1 = xind0 & ((np.nanmedian(D1, axis=1) > thresh)[:, None]) & (
    (np.nanmedian(D2, axis=1) > thresh)[:, None]) & np.isfinite(X)


def MAD(x):
    return np.nanmedian(np.abs(x - np.nanmedian(x)))


bins = 50

S = scipy.stats.binned_statistic(X[xind1],
                                 Y[xind1],
                                 MAD,
                                 range=[0, 2],
                                 bins=bins)
SC = scipy.stats.binned_statistic(X[xind1],
                                  Y[xind1],
                                  'count',
                                  range=[0, 2],
                                  bins=bins)

plt.clf()
aind = SC.statistic > 100
plt.plot((S.bin_edges[:-1] + .5 * np.diff(S.bin_edges))[aind],
         S.statistic[aind] * 1.48, 'o')
plt.ylim(0, 3)
plt.xlim(0.3, 2)
plt.xlabel('log10(SN)')
plt.ylabel('Stddev/sigma')
```
